pandas/core/indexing.py

```diff
@@ -1703,6 +1703,10 @@ def _setitem_with_indexer(self, indexer, value, name: str = "iloc"):
         # maybe partial set
         take_split_path = not self.obj._mgr.is_single_block
 
+        if not take_split_path and isinstance(value, ABCDataFrame):
+            # Avoid cast of values
+            take_split_path = not value._mgr.is_single_block
```

Comment:
no good alternatives?  recall that with the recently-enforced deprecation the idea is to get rid of the two-paths altogether

Reply:
could use frame._is_homogeneous_type

Reply:
Not in the short term without duplicating lots of stuff. The indexer is prepared differently, which causes all sorts of problems and complications.

If we want to get rid of both paths, we have to unify this anyway. This wouldn't hurt in this case.